### PR TITLE
Transition the U Utah SLR algorithm to UPP

### LINKED ISSUE #662: Transition the U Utah SLR algorithm to UPP
Transition the U Utah SLR algorithm to UPP with PR https://github.com/NOAA-EMC/UPP/pull/661

### PR DESCRIPTION
**Pull request by JesseMeng-NOAA**
Code for calculating SLR using the U Utah algorithm is developed in UPP_PHYSICS.f module.
Currently MDL2P.f calls calslr_roebber() to calculate SLR and outputs the grib2 variable SDEN (SLR=1000./SDEN), which is already in UPP/develop and adopted by parallel rrfs.
This PR adds the option to call calslr_uutah(), in a commented line for future implementation.  
          call calslr_roebber(TPRS,RHPRS,EGRID1)
      !   call calslr_uutah(EGRID1)

UPP issue https://github.com/NOAA-EMC/UPP/issues/662

### CODE REVIEW

**Comment by WenMeng-NOAA** [comment]
@JesseMeng-NOAA Do you have an UPP issue for this PR? If not, please open one.

**Comment by JesseMeng-NOAA** [comment]
> @JesseMeng-NOAA Do you have an UPP issue for this PR? If not, please open one.

https://github.com/NOAA-EMC/UPP/issues/566
was originally for transition TWO SLR algorithms but I noticed you closed it after we merged Roebber..
Do you want to reopen #566 or have a new issue?

**Comment by WenMeng-NOAA** [comment]
From the code management perspective, it would be better you create a new one to link this PR. @JesseMeng-NOAA

**Comment by FernandoAndrade-NOAA** [comment]
Hi, @JesseMeng-NOAA UPP regression tests have completed on Hera and Orion with no changes to results.

**Comment by JesseMeng-NOAA** [comment]
> Hi, @JesseMeng-NOAA UPP regression tests have completed on Hera and Orion with no changes to results.

Thank you @FernandoAndrade-NOAA !

**Comment by WenMeng-NOAA** [comment]
@JesseMeng-NOAA Can you sync your branch with upstream/develop?

**Comment by JesseMeng-NOAA** [comment]
> @JesseMeng-NOAA Can you sync your branch with upstream/develop?
@WenMeng-NOAA 
Synced

File: sorc/ncep_post.fd/MDL2P.f [hunk 1]
```diff
@@ -4082,13 +4083,18 @@ SUBROUTINE MDL2P(iostatusD3D)
...
             if(egrid1(i,j) < spval) then
-                grid1(i,j)=1000./egrid1(i,j)
-            else
-                grid1(i,j)=spval
+              if(egrid1(i,j)>=1.) then
+                 grid1(i,j)=1000./egrid1(i,j)
+              else
+                 grid1(i,j)=spval
+              endif
+            !  grid1(i,j)=egrid1(i,j)
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/MDL2P.f]: @JesseMeng-NOAA Is this line for UTA SLR output?

File: sorc/ncep_post.fd/MDL2P.f [hunk 2]
```diff
@@ -4082,13 +4083,18 @@ SUBROUTINE MDL2P(iostatusD3D)
          if(me==0)PRINT*,'CALLING SLR'
          egrid1=spval
          call calslr_roebber(TPRS,RHPRS,EGRID1)
```
Comment (JesseMeng-NOAA) [inline sorc/ncep_post.fd/MDL2P.f]: The SLR Roebber / Utah switch is made here.
Default is Roebber.

File: sorc/ncep_post.fd/MDL2P.f (hunk 1 above)
Comment (WenMeng-NOAA) [reply sorc/ncep_post.fd/MDL2P.f]: @JesseMeng-NOAA I would like to run a testing for UTA SLR. The UTA SLR should be assigned grid1=egrid1 or grid1=1000./egrid1?

Comment (JesseMeng-NOAA) [reply sorc/ncep_post.fd/MDL2P.f]: The SLR Roebber / Utah switch is made here.
MDL2P.f
4085         call calslr_roebber(TPRS,RHPRS,EGRID1)
4086      !   call calslr_uutah(EGRID1)
Default is Roebber.

The line you referred is a debugging line. We calculate SLR but grib2 outputs SDEN which is 1000/SLR and it is easier to debug by outputting the original values. Can be removed if you prefer.

Comment (WenMeng-NOAA) [reply sorc/ncep_post.fd/MDL2P.f]: @JesseMeng-NOAA Got it. Thanks!

**Comment by WenMeng-NOAA** [comment]
@JesseMeng-NOAA I ran testing to generate Utah SLR on dogwood. You may find results at:
GFS: /u/wen.meng/ptmp/gfs.20190830
RRFS: /u/wen.meng/ptmp/fv3r_2022030200
Please let me know if you see anything inconsistent with your testing.

**Comment by JesseMeng-NOAA** [comment]
> @JesseMeng-NOAA I ran testing to generate Utah SLR on dogwood. You may find results at: GFS: /u/wen.meng/ptmp/gfs.20190830 RRFS: /u/wen.meng/ptmp/fv3r_2022030200 Please let me know if you see anything inconsistent with your testing.

OK I will take a look.

File: sorc/ncep_post.fd/MDL2P.f [hunk 3]
```diff
@@ -4082,13 +4083,18 @@ SUBROUTINE MDL2P(iostatusD3D)
...
+      !   call calslr_uutah(EGRID1)
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/MDL2P.f]: @JesseMeng-NOAA I wonder if you can use a flag to switch between two schemes. This flag can be read from itag as namelist. The default setting is Roebber scheme. The sample flag can be found in UPP code such as "gocart_on"

**Comment by JesseMeng-NOAA** [comment]
> > @JesseMeng-NOAA I ran testing to generate Utah SLR on dogwood. You may find results at: GFS: /u/wen.meng/ptmp/gfs.20190830 RRFS: /u/wen.meng/ptmp/fv3r_2022030200 Please let me know if you see anything inconsistent with your testing.
> 
> OK I will take a look.

Where are your input files?

**Comment by WenMeng-NOAA** [comment]
> > > @JesseMeng-NOAA I ran testing to generate Utah SLR on dogwood. You may find results at: GFS: /u/wen.meng/ptmp/gfs.20190830 RRFS: /u/wen.meng/ptmp/fv3r_2022030200 Please let me know if you see anything inconsistent with your testing.
> > 
> > 
> > OK I will take a look.
> 
> Where are your input files?

@JesseMeng-NOAA I use model files for gfs and rrfs from the UPP RT tests at /u/wen.meng/noscrub/ncep_post/post_regression_test_new/data_in:
gfs: gfs/
rrfs: fv3r/

**Comment by JesseMeng-NOAA** [comment]
> > > > @JesseMeng-NOAA I ran testing to generate Utah SLR on dogwood. You may find results at: GFS: /u/wen.meng/ptmp/gfs.20190830 RRFS: /u/wen.meng/ptmp/fv3r_2022030200 Please let me know if you see anything inconsistent with your testing.
> > > 
> > > 
> > > OK I will take a look.
> > 
> > 
> > Where are your input files?
> 
> @JesseMeng-NOAA I use model files for gfs and rrfs from the UPP RT tests at /u/wen.meng/noscrub/ncep_post/post_regression_test_new/data_in: gfs: gfs/ rrfs: fv3r/

Your results are identical to mine. The output SLR fields are reasonable.

**Comment by WenMeng-NOAA** [comment]
@JesseMeng-NOAA Thanks for verifying.

File: sorc/ncep_post.fd/MDL2P.f (hunk 3 above)
Comment (JesseMeng-NOAA) [reply sorc/ncep_post.fd/MDL2P.f]: OK I can do that.

**Comment by WenMeng-NOAA** [comment]
@JesseMeng-NOAA Can you sync your branch with the latest UPP develop again?

**Comment by WenMeng-NOAA** [comment]
@HuiyaChuang-NOAA Please let me know if you have any comments on this PR. I will proceed to the final testing.

**Comment by JesseMeng-NOAA** [comment]
> @JesseMeng-NOAA Can you sync your branch with the latest UPP develop again?

synced

**Review by WenMeng-NOAA — APPROVED** [review]

**Comment by WenMeng-NOAA** [comment]
The UPP RT tests were completed on WCOSS2, Hera and Orion. No baseline update is needed with this PR.

**Pull request merged**